src/EVM/SMT.hs

```diff
@@ -210,7 +210,13 @@ assertProps :: Config -> [Prop] -> SMT2
 assertProps conf ps = assertPropsNoSimp conf (decompose . Expr.simplifyProps $ ps)
   where
     decompose :: [Prop] -> [Prop]
-    decompose props = fromMaybe props (mapM (mapPropM Expr.decomposeStorage) props)
+    decompose props = if conf.decomposeStorage && (isJust $ sequence_ safeList1) && (isJust $ sequence_ safeList2)
+                      then fromMaybe props (mapM (mapPropM Expr.decomposeStorage) props)
+                      else props
+      where
+        -- All in these lists must be a `Just ()` or we cannot decompose
+        safeList1 = map (mapPropM_ Expr.safeToDecompose) props
```

Comment:
`all $ map (foldProp Expr.safeToDecompose) props` would have the same effect as 
`(isJust $ sequence_ (map (mapPropM_ Expr.safeToDecompose) props))` without the use of monads

Reply:
Ah nice! I didn't know about the `all` predicate! So much cleaner, thank you!